Added functionality for Incident

excelciordan opened the pull request:
Two pieces of valuable information are not included with an incdient:

1. Whether the summary_details includes a subject or a description
2. The extra data contained in the details log entry

With this small code update, an incident is now aware of both of these items.  There are two new methods for incidents:

1. incident.get_detail_log_entry(requester_id)  #Returns json of log entry
2. incident.has_subject()  # Boolean return, if no subject, than it has details (incident.trigger_summary_data.subject/details)

gmjosack commented:
I haven't taken a look at this too much yet but before I do can you do the following:

* Fix the flake errors that are causing the build to fail
* Remove debug printing
* Remove the `.idea/dictionaries/dguimaraes.xml` file from the commit and squash all commits into a single commit. (Possibly add `.idea` to the `.gitignore` to prevent those making it in in the future)

Thanks!

excelciordan commented:
I made a couple of the changes you requested.  Still trying to figure out how to squash the commits.  Also, Trying to understand the error that is occuring so I can fix it.

gmjosack commented:
I can handle the squash when I merge so don't worry about it.

gmjosack commented on pygerduty/__init__.py:
```diff
@@ -391,11 +395,20 @@ def __init__(self, *args, **kwargs):
+        try:
+            self.trigger_summary_data.subject  # try to access the subject
+            has_subject = True
+        except:
```
In general I like to prevent naked exceptions. I imagine you're looking to catch an `AttributeError` when the `subject` doesn't exist on `self.trigger_summary_data`. Could this be simplified to 

```
def has_subject(self):
    return hasattr(this.trigger_summary_data, "subject")
```

excelciordan commented:
Yes!!  That is much simpler.  I write a lot of ruby and maintain a lot of perl.  I was looking for a better way to do this.  Thank you!  I'll commit and submit the change.

gmjosack commented on pygerduty/__init__.py:
```diff
@@ -403,6 +416,11 @@ def resolve(self, requester_id):
     def acknowledge(self, requester_id):
         self._do_action('acknowledge', requester_id=requester_id)
 
+    def get_details_log_entry(self, requester_id):
```
I'm not certain about this method name. This specifically grabs the single log entry related to triggering the incident. You can currently list all log_entries with the channel like so

`incident.log_entries.list(include=["channel"])`

If you're looking specifically for the "trigger_log_entry" it's worth being more explicit with the name. If that's the case then you can use the id from `trigger_details_html_url` with `self.log_entries.show(some_id, include=["channel"])` instead of modifying the `_do_action` method. That should further simplify your actual pull request.

gmjosack replied:
Yeah, thinking about it more I think it would make the most sense to have something like

```
def get_trigger_log_entry(self, **kwargs):
    match = TRIGGER_DETAILS_ID_RE.search(self.trigger_details_html_url)
    return self.log_entries.show(match.group("trigger_id"), **kwargs)
```

This way you get easy access to the triggering log event and still have control over the specific options you pass through.

Would that meet your needs?

excelciordan commented:
This suggestion helped me to understand your code a lot better.  I see how that works and I implemented the changes.  I wanted your input on the name of the method before I make another commit.  I dropped the word get from the method name so it is just called details_log_entry

i.e. incident.details_log_entry()

Your solution was more elegant.  The requester_id had no business being there nor did making the extra call to the PD API.  Thank you for taking the time to step though this with me!

gmjosack commented:
Yeah, when I initially wrote this library it was made, possibly too, generic to adapt to new options being added to the API without needing code changes. The result being the code is a bit complex to wrap your head around at first and the lack of docs doesn't help.

No problem!

excelciordan commented:
I really like generic code and I'm always striving to generalize mine.  I'm not in a development role so I don't have a lot of support in that area (even though I spend most of my time developing).  I really appreciate this opportunity to walk though this with you.

gmjosack commented:
Don't know if you missed this comment because I double commented on the same line but reiterating it here:

> Yeah, thinking about it more I think it would make the most sense to have something like

```python
def get_trigger_log_entry(self, **kwargs):
    match = TRIGGER_DETAILS_ID_RE.search(self.trigger_details_html_url)
    return self.log_entries.show(match.group("trigger_id"), **kwargs)
```

> This way you get easy access to the triggering log event and still have control over the specific options you pass through.
>
> Would that meet your needs?

Specifically an incident has many log entries. This method is asking about a specific log entry, the one that triggered the incident. I do see the value is having easy access to the log entry that triggered the incident but I'm not sure it's worth having a method that always returns the detailed output opposed to just allowing that to be passed through. For example with the method above you could just call `incident.get_trigger_log_entry(include=["channel"])`

excelciordan commented:
Sorry, I did miss that.  I fixed it.

gmjosack commented:
I manually squashed the commits when merging so this won't close cleanly. You'll want to fix up your local fork or refork before making future contributions.

I'll close this manually. Thanks for the contribution! This has been pushed to PyPI as pygerduty-0.32.0
